File: ProyectoFinalEDD/source/Graph.cpp
#include "../headers/Graph.h"

void Graph::addEdge(int u, int v) {

    adjList[u].push_back(v);

    adjList[v].push_back(u);
}

void Graph::display() {

    cout << "\n===== GRAFO =====\n";

    for(int i = 0; i < SIZE; i++) {

        cout << i << " -> ";

        for(int neighbor : adjList[i]) {

            cout << neighbor << " ";
        }

        cout << endl;
    }
}

void Graph::BFS(int start) {

    bool visited[SIZE] = {false};

    queue<int> q;

    visited[start] = true;

    q.push(start);

    cout << "\n===== BFS =====\n";

    while(!q.empty()) {

        int vertex = q.front();

        q.pop();

        cout << vertex << " ";

        for(int neighbor
            : adjList[vertex]) {

            if(!visited[neighbor]) {

                visited[neighbor] = true;

                q.push(neighbor);
            }
        }
    }

    cout << endl;
}

void Graph::DFSUtil(int vertex,
                    bool visited[]) {

    visited[vertex] = true;

    cout << vertex << " ";

    for(int neighbor
        : adjList[vertex]) {

        if(!visited[neighbor]) {

            DFSUtil(neighbor,
                    visited);
        }
    }
}

void Graph::DFS(int start) {

    bool visited[SIZE] = {false};

    cout << "\n===== DFS =====\n";

    DFSUtil(start,
            visited);

    cout << endl;
}

File: ProyectoFinalEDD/java/src/Graph.java
import java.util.*;

public class Graph {

    private Map<Integer,
    List<Integer>> adjList;

    public Graph() {

        adjList =
        new HashMap<>();
    }

    public void addEdge(int u,
                        int v) {

        adjList.putIfAbsent(
        u,
        new ArrayList<>());

        adjList.putIfAbsent(
        v,
        new ArrayList<>());

        adjList.get(u).add(v);

        adjList.get(v).add(u);
    }

    public void display() {

        System.out.println(
        "\n===== GRAFO =====");

        for(Integer node :
            adjList.keySet()) {

            System.out.print(
            node + " -> ");

            for(Integer neighbor :
                adjList.get(node)) {

                System.out.print(
                neighbor + " ");
            }

            System.out.println();
        }
    }

    public void bfs(int start) {

        System.out.println(
        "\n===== BFS =====");

        Set<Integer> visited =
        new HashSet<>();

        Queue<Integer> queue =
        new LinkedList<>();

        visited.add(start);

        queue.add(start);

        while(!queue.isEmpty()) {

            int current =
            queue.poll();

            System.out.print(
            current + " ");

            for(Integer neighbor :
                adjList.get(current)) {

                if(!visited.contains(
                   neighbor)) {

                    visited.add(
                    neighbor);

                    queue.add(
                    neighbor);
                }
            }
        }

        System.out.println();
    }

    public void dfs(int start) {

        System.out.println(
        "\n===== DFS =====");

        Set<Integer> visited =
        new HashSet<>();

        dfsRecursive(start,
                     visited);

        System.out.println();
    }

    private void dfsRecursive(
            int node,
            Set<Integer> visited) {

        visited.add(node);

        System.out.print(
        node + " ");

        for(Integer neighbor :
            adjList.get(node)) {

            if(!visited.contains(
               neighbor)) {

                dfsRecursive(
                neighbor,
                visited);
            }
        }
    }
}